Shreya P taking over for Abhiram Bylahalli Jagadish – Educator Task Submissions UI Implementation

abhirambj opened the pull request:
# Description

This PR implements a modern, responsive UI for educators to view and manage student task submissions in the education portal. The interface allows educators to view all submissions organized by classes and tasks, with filtering capabilities and detailed submission information.

**Implements:** Feature request for "Display List of Finished Tasks Reported by Students"

## Related PRs (if any):

To test this frontend PR, you need to check out the backend PR from HGNRest.

Backend PR: https://github.com/OneCommunityGlobal/HGNRest/pull/1799
Figma Design: https://www.figma.com/design/vpde7g6et8Xl2pfyp5lg1i/HGN-Phase-4--Highest-Good-Education?node-id=1435-2019

---

## Main changes explained:

### New Components Created:
- **TaskSubmissionsPage.jsx** - Main container component with class tabs, task sections, and filtering
- **SubmissionCard.jsx** - Individual student submission card with status badges and grade display
- **TaskSubmissionsPage.module.css** - Responsive styling with pill-style tabs and modern layout
- **SubmissionCard.module.css** - Card styling with gradients, tooltips, and status indicators

### Key Features Implemented:
- **Class-based tab navigation** with horizontal scrolling and arrow buttons for overflow
- **Expandable task sections** showing submission counts and due dates
- **Filter dropdown** for viewing all submissions, pending reviews, or graded work
- **Student cards** with:
  - Dynamically colored avatars generated from student initials
  - Status badges (green "✓ Graded" badge, info icons for pending)
  - Grade display sections for completed work
  - Late submission warnings with red borders and timestamps
  - White tooltips showing submission details
- **Responsive design** adapting from multi-column grid on desktop to single column on mobile
- **Smooth animations** with hover effects and transitions (0.2s ease)
- **Accessibility features** including keyboard navigation, ARIA labels, and focus states

### Data Management:
- Integration with backend API endpoint `/api/educationportal/educator/task-submissions`
- Intelligent grouping by lesson plan (classes) and task names using React useMemo
- Real-time filtering without page reloads
- Proper error handling and loading states

### UI/UX Improvements:
- Modern pill-style tabs matching Figma specifications
- Subtle gray section headers with smooth hover states
- Professional color scheme with gradients for visual hierarchy
- Consistent 8px grid spacing system
- Touch-friendly 44px minimum touch targets on mobile

---

## How to test:

### Prerequisites:
1. Checkout and run the backend PR [#1799](https://github.com/OneCommunityGlobal/HGNRest/pull/1799) from HGNRest repository
2. Start the frontend development server and connect with backend port.

### Testing Steps:

**1. Login and Navigation:**
- Login with an educator account
- Navigate to `/educationportal/educator/task-submissions`
- Verify the page loads with submission data

**2. Class Tab Navigation:**
- Click different class tabs to switch between lesson plans
- Verify horizontal scrolling works when many classes exist
- Test arrow buttons for tab navigation
- Confirm active tab has blue gradient and elevation

**3. Filter Functionality:**
- Use the "Select student list" dropdown
- Test "All Submissions" - should show everything
- Test "Submissions pending" - should show only pending reviews
- Test "Submissions received" (Graded) - should show only graded submissions
- Verify submission counts update based on filters

**4. Task Section Interactions:**
- Click task section headers to expand/collapse
- Verify submission count is displayed correctly
- Check due date formatting (e.g., "Due Oct 18 at 7:00 PM")
- Test keyboard navigation (Tab, Enter, Space keys)

**5. Submission Cards:**
- **Pending submissions**: Should show info icon in top-right
  - Hover over info icon to see white tooltip with "Pending review"
- **Graded submissions**: Should show green "Graded" badge
  - Should display grade value (A, B, C, etc.) at bottom
  - Should have light green background gradient
- **Late submissions**: Should have red border and warning section
  - Hover over info icon to see tooltip with late submission details
  - Should show "LATE SUBMISSION" text with timestamp
- Verify student names display correctly with colored avatars
- Check task type labels (Read Task, Write Task, Quiz Task, etc.)

**6. Responsive Design:**
- Test on desktop (1920px, 1440px, 1024px widths)
- Test on tablet (768px width)
- Test on mobile (375px, 414px widths)
- Verify cards stack properly on smaller screens
- Check touch scrolling works on mobile tabs

**7. Edge Cases:**
- Test with no submissions (should show "No submissions found")
- Test with filtered results returning empty (should show appropriate message)
- Test with single submission vs many submissions
- Verify loading state shows spinner
- Test error state (disconnect backend and verify error message)

---

## Screenshots:

### Desktop View - All Submissions

<img width="1505" height="909" alt="image" src="https://github.com/user-attachments/assets/a1458466-e385-4e6a-8340-bb11c8a5ac2a" />

### Graded Submission Card

<img width="1508" height="905" alt="image" src="https://github.com/user-attachments/assets/7882cec1-e3bf-40d5-a8c5-ccf0dd5590e6" />

### Late Submission Card
<img width="1512" height="909" alt="image" src="https://github.com/user-attachments/assets/a75dc13e-2936-4488-8cb4-fd5ab48a649b" />

### Mobile Responsive View
<img width="1506" height="907" alt="image" src="https://github.com/user-attachments/assets/12d29070-dcb1-49c4-91f0-41205901490b" />

### Tooltip Interaction
<img width="380" height="291" alt="image" src="https://github.com/user-attachments/assets/df5aebba-3dca-4900-935b-c04eedf4e295" />

---

abhirambj committed:
feat: Implement Task Submissions Page with Submission Cards

- Added SubmissionCard component to display individual submission details including student name, avatar, submission status, and timestamps.
- Created TaskSubmissionsPage component to fetch and display grouped submissions by class and task, with filtering options for submission status.
- Introduced CSS modules for styling SubmissionCard and TaskSubmissionsPage, ensuring a responsive and user-friendly interface.
- Implemented loading and error states for better user experience during data fetching.
- Added functionality to expand/collapse task sections to view submissions.

abhirambj committed:
feat: revamp educator task submissions UI with modern design

- Create TaskSubmissionsPage with class tabs and task sections
- Implement SubmissionCard component with status badges and grades
- Add pill-style tabs with horizontal scroll and arrow navigation
- Build filter dropdown for all/pending/graded submissions
- Design responsive cards with colored avatars and tooltips
- Add late submission warnings with red borders
- Implement expandable sections with submission counts
- Match Figma design specifications throughout

Sriamshreddy000 commented:
Hi Abhiram,
Tested this PR along with the related backend [[PR #1799]](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/1799). After navigating and logging in, the page fails to load submissions and displays the message “Failed to load submissions. Please try again.” In the browser console, I’m seeing a 404 (Not Found) error.Not sure if this might be a setup issue on my end or something related to the PR itself, but I wanted to point it out just in case.

I’m attaching a screenshot below for reference.
<img width="1920" height="1038" alt="Screenshot 2025-11-11 at 8 08 42 PM" src="https://github.com/user-attachments/assets/f296fee4-fd96-4e21-839b-a17ba15433dc" />

roshini1212 reviewed (commented):
Tested the PR and encountered this error. Unsure whether it's a local setup issue or PR-related, but including the screenshot for reference.

<img width="1512" height="592" alt="Image" src="https://github.com/user-attachments/assets/2f7c1bf0-0e3e-456a-874e-ddce3870a189" />

ShreyaMP1999 committed:
fix(education): improve educator task submissions UI stability and data handling

- fix duplicate SubmissionCard rendering issue
- ensure stable React keys for submission cards
- normalize status filtering for consistency with backend
- add date validation to prevent invalid comparisons
- improve avatar initials handling for edge cases

ShreyaMP1999 reviewed (dismissed):
https://www.dropbox.com/scl/fo/3uqzokj7v8mtktqb7i79i/AAb9tkmADXGhsreZjMQShL8?rlkey=e5vpdhilt1jso0ak9fhyd4ol8&st=qfxmrbce&dl=0

I reviewed the existing implementation for the Educator Task Submissions feature and made a few necessary improvements to both the frontend and backend to ensure stability and consistency. On the frontend, I fixed duplicate rendering issues, ensured stable React keys for submission cards, normalized status handling to align with backend responses, and added safer date handling to avoid invalid comparisons. I also improved edge-case handling for student name initials and ensured better resilience in UI rendering. On the backend, I verified the response structure and filtering logic to confirm it aligns correctly with the frontend requirements.

I tested the complete flow end-to-end, including API validation in Postman, UI behavior in the browser, and responsiveness using Safari mobile simulator. All core functionalities such as filtering, grouping, status display, and late submission detection are working as expected. Screenshots and testing evidence have been uploaded to the Dropbox link for reference.

Anusha-Gali reviewed (dismissed):
Hi Shreya,

I have reviewed your PR locally and it works as per requirement.

<img width="1278" height="824" alt="Image" src="https://github.com/user-attachments/assets/e8f6256a-81d4-44ac-bf7e-c1cd9e38cade" />
<img width="1278" height="824" alt="Image" src="https://github.com/user-attachments/assets/f79617a1-6012-4f9e-a7ca-57779a369b62" />
<img width="1278" height="824" alt="Image" src="https://github.com/user-attachments/assets/387c8400-4d0a-4684-a8be-2aa52d567dfd" />
<img width="1278" height="824" alt="Image" src="https://github.com/user-attachments/assets/c5babbf9-b2d9-4dbd-b03b-921d928e92a0" />
<img width="1278" height="824" alt="Image" src="https://github.com/user-attachments/assets/bf994c2d-573b-4662-956e-c3c9b4d20f8a" />

ShreyaMP1999 committed:
fix: resolve merge conflicts and align with development

SharadhaKasiviswanathan commented:
Tested the Student Dashboard flow where the page loads, but it is showing repeated Failed to fetch sub-tasks errors on load. The intermediate-task URLs here are pointing to the /educator/... path, while the backend mounts intermediate-task routes under /api/intermediate-tasks/.... This flow could be checked if the frontend endpoint is using the wrong base path, because it is causing the sub-task requests to fail in the dashboard.
<img width="1902" height="864" alt="image" src="https://github.com/user-attachments/assets/0f39aea7-656f-4e2c-97e3-bbd78f56feb7" />

SharadhaKasiviswanathan commented on src/components/EductionPortal/Educators/TaskSubmissions/TaskSubmissionsPage.jsx:
```diff
@@ -0,0 +1,270 @@
+    const fetchSubmissions = async () => {
+      try {
+        setLoading(true);
+        const res = await axios.get(
```
Tested the new Educator Task Submissions flow from the frontend side, the page and route are added properly, but the API call is pointing to /educationportal/educator/task-submissions and I do not see a matching backend route mounted in HGNRest. The backend here still exposes the education tasks routes and the educator routes separately, but not this submissions endpoint. This submissions flow could be checked if the backend API is missing in this PR or if the frontend is currently wired to a route which isn't implemented.

sundarmachani reviewed (approved):
I tested this PR, and everything is working as expected. Also, test the PR linked with the Backend one; it functions and looks good.

<img width="1728" height="1039" alt="Image" src="https://github.com/user-attachments/assets/5b88ae5a-9d7c-449e-a941-e0f45fdc9f25" />

one-community commented:
Thank you all, merging!
